src/main/java/net/airvantage/sched/services/tech/hystrix/WakeupCommand.java

```
@@ -0,0 +1,105 @@
+package net.airvantage.sched.services.tech.hystrix;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.hystrix.HystrixCommand;
+import com.netflix.hystrix.HystrixCommandGroupKey;
+import com.netflix.hystrix.HystrixCommandProperties;
+import com.netflix.hystrix.HystrixThreadPoolProperties;
+
+import net.airvantage.sched.app.exceptions.ServiceRuntimeException;
+import net.airvantage.sched.model.JobWakeup;
+import net.airvantage.sched.quartz.job.JobResult;
+import net.airvantage.sched.quartz.job.JobResult.CallbackStatus;
+import net.airvantage.sched.services.tech.JobExecutionHelper;
+import net.airvantage.sched.services.tech.RetryPolicyHelper;
+
+public class WakeupCommand extends HystrixCommand<WakeupResult> {
+    private static final Logger LOG = LoggerFactory.getLogger(WakeupCommand.class);
+
+    public static final String DEFAULT_GROUP = "defaultGroup";
+    public static final int QUEUE_SIZE = 2_000;
+
+    private final JobWakeup wakeup;
+    private final JobExecutionHelper jobExecutionHelper;
+    private final RetryPolicyHelper retryPolicyHelper;
+    private JobResult jobResult;
+
+    public WakeupCommand(JobWakeup wakeup, JobExecutionHelper jobExecutionHelper, RetryPolicyHelper retryPolicyHelper,
+            int threadPoolSize) {
+        super(Setter.withGroupKey(makeGroup(wakeup))
+                .andThreadPoolPropertiesDefaults(HystrixThreadPoolProperties.Setter()
+                        .withQueueSizeRejectionThreshold(QUEUE_SIZE)
+                        .withMaxQueueSize(QUEUE_SIZE)
+                        .withCoreSize(threadPoolSize))
+                .andCommandPropertiesDefaults(HystrixCommandProperties.Setter()
+                        .withExecutionTimeoutEnabled(false)));
```

By default Hystrix manages a timeout of 1 sec. It seems to be too small for AvCore services. I think 30 seconds should be better. Otherwise we will have a very high error level.

---

@sbuffo - I disabled the timeout.  The existing code in `jobExecutionHelper.execute(wakeup)`  (and other helpers) manages the timeouts for these tasks already and I didn't want to interrupt this processing since it looks like it handles multiple retries.

maybe we could enable a timeout at the hystrix level just in case something happens?  what would a good value be 30 seconds?  It looks like the current http execution can retry multiple times each time with a 60 second timeout.

---

I don't think that the `RetryPolicyHelper` actually retries http calls. It updates the job in DB with a new retry date.
If it is true we could interrupt the execution and let the fallback code di the same job update logic.

---

Indeed, I did not see that you have disabled the timeout :-).
But I understood that a reason to introduce Hystrix was to manage query timeout ( and also to short-circuit it if there are a lot of timeouts ). So I would enable the Hystrix command timeout and update the current retry policy to not exceed the configured value. I think 30 sec it is a good value for a first step.
It would be usefull to monitor it ( error/timeout rate ) and validate it is an efficient configuration. We already uses "Codahale Metrics" ( e.g into JobSchedulingServiceImpl ), so maybe we can reuse it.

---

@rdouyere / @sbuffo  - this  https://github.com/AirVantage/av-sched/blob/PLTFRS-9638_wakeup_bulkhead/src/main/java/net/airvantage/sched/services/tech/RemoteServiceConnector.java#L53-L98

is configured to retry 7 times for each http request made.  each one has a 60 second timeout. this is shared between the cron jobs and wakeups.

it seems that we shouldn't have this retry at this level and let the jobs get rescheduled if they fail once?    @sbuffo  is this what you are suggesting to change?

---

Thanks @bluejeff31  for the pointer on the retry code. 7 * 60 seems pretty long => during this time the whole loop is waiting if a single wakeup is retried the maximum number. This could be aggravated if the http call hangs.

---

RemoteServiceConnector is usefull to manage the basic connection errors ( HTTP 502 / 503, IOException ) because most of time short retries allow to manage this kind of connection errors. But if the http request takes too much time for any reason then the Hystrix timeout will stop it. So just add an Hystrix timeout should be enough. @bluejeff31 @rdouyere What do you think ?

---

Would agree to add an Hystrix timeout but only if RemoteServiceConnector is really limited to quick retries. Like up to 10 seconds total.

---

@rdouyere - I agree with making the RemoteServiceConnector retries shorter (10 s) + adding an overall hystrix timeout (30 s) .  This will improve failing faster.

my only concern is are there existing wakeups that take longer than 10 seconds to execute?  I hope not.

---

In my mind 10 seconds are time when calculating retries. If we initiate an HTTP call it won't cut after 10s. Actually the HTTP client seems to be configured to have a 60 seconds connection timeout & a 60 seconds request timeout.

---

@rdouyere / @sbuffo  - looking closer at this - the `RemoteServiceConnector`  is configured to do 7 retries and backs off using this: `return (long) Math.pow(2, retryCount) * 100L;`   so from 100 ms  to about 13 s (at the 7th try).  Each of these tries could take up to 60 s.

I suggest we set hystrix timeout to 60 s and cut down the 7 retries to 2 or 3

* this will more quickly fail cases when the callback is not responding quickly so we won't end up waiting X * 60 s max wait will be 60 s - if a system is this slow rescheduling is probably better

* it still allows the quick retries to happen

---

👍

---

@rdouyere / @sbuffo  - I configured as suggested above (2 retries at http level and 60 s hystrix timeout).  one behavior may not be as desired:  the hystrix timeout takes effect from the time the task is queued not when it is actually started.   In the case the callback is very long and lots of requests are pending (I tested with 10 seconds)  
* the queued tasks will all start to fail after 60 seconds
* this triggers a circuit breaker to open 
* and reschedules all the failed tasks.

with a consistently slow callback and large work queue the rescheduling will delay tasks.

I will see if the timeout case can be detected and rescheduled tasks retry_count will not be incremented.  This might improve the situation.

on the other hand the bulkhead pattern is working, The bulk task selection logic from the database might need improvement so a large amount of tasks for one component can't starve the others.

---

I reverted back to no hystrix timeout and the code behaves better when there are long latencies and large backups.  I see a few options here:

1 rely on the http helper to timeout and hystrix will not enforce this - how the code currently works
2 in addition to above - make the DB selection logic smart enough passing in groupnames that have full queues to avoid selecting more tasks for these.  I'm not sure hystrix will let us get to this information.

At this point I think doing #1 is safe enough.  we will see a benefit with the bulkhead

---

@bluejeff31 for your last point, instead of a single query loop on DB to get wakeup to process, we could also have as many concurrent loops as different group names. Each would read & queue wakeups independently. With a blocking queue doing backpressure.

---

@rdouyere - I agree multiple db selectors one for each group would be best.  I'll look at the DB selection but perhaps just adding the bulkhead is enough for now?

---

yep, is it ok for a final review then?

---

@rdouyere - just pushed the changes mentioned above - it's ready.